### ISSUE TITLE
e2e tests can be silently skipped and look like they are passing

### ISSUE THREAD

**Issue by jlowdermilk** [kubernetes/kubernetes#25515]
Filed from #25426. A conformance test was failing on the 1.2 branch that should actually have been failing everywhere. The test itself is broken (not the functionality, thankfully), but wasn't doing anything because `forEachPod` was looping over an empty set [source]:(https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl.go#L861):
```
...
		It("should be able to retrieve and filter logs [Conformance]", func() {
			framework.SkipUnlessServerVersionGTE(extendedPodLogFilterVersion, c)

-->			forEachPod(func(pod api.Pod) {
				By("checking for a matching strings")
				_, err := framework.LookForStringInLog(ns, pod.Name, containerName, "The server is now ready to accept connections", framework.PodStartTimeout)
				Expect(err).NotTo(HaveOccurred())

...
```

TODO:
- Fix `forEachPod` to fail if the list is empty. 
- Audit tests and make sure tests that claim to be passing are actually doing something.

**Comment by jlowdermilk** [comment]
Ok, the skip isn't actually broken for #25426, it's just running a for loop over an empty list. Probably still worth doing an audit to make sure there aren't other spuriously passing tests.

**Comment by jayunit100** [comment]
FWIW 
- `clusterState.ForEachPod` cannot/should not be fixed - it is an explicti cluster operation.  
- `forEachPod` is just a wrapper used in this particular test.
- If you want to fail if no pods conform to a particular state `clusterState.WaitFor(1,1*time.Minute)` which will fail if < 1 pod exists after a minute.

**Comment by jayunit100** [comment]
And also I totally agree we should phase out the `skipIf` stuff b/c it can yield false positives iirc

**Comment by jlowdermilk** [comment]
IMO, `clusterState.ForEachPod` *should* fail by default if no pods match the filter. Otherwise, the test looks like it's passing but the check function is never run. I made that change in #25537. If you object to it, please let me know.

**Comment by jayunit100** [comment]
Clusterstate.WaitFor(1) followed by Clisterstate.ForEach accomplishes what you need explicitly without adding too much extra boilerplate, right?

**Comment by jlowdermilk** [comment]
ClusterState.Waitfor(1) does the trick, and is the change I made, but ForEachPod still needs to fail if the filter is empty. It was not obvious that creating an rc then immediately calling `ForEachPod` would be a no-op. Anyway this can be closed once #25537 is cherry-picked onto 1.2; it fixed the two tests that weren't calling `WaitFor` before `ForEachPod`.

**Comment by jayunit100** [comment]
Haha ok I guess the for each semantics have already drifted so it's too late :).

**Comment by jlowdermilk** [comment]
Fixed by #25426
